File: doc-examples/jdbc-example-kotlin/src/main/kotlin/example/SaleRepository.kt

package example

import io.micronaut.data.annotation.Join
import io.micronaut.data.annotation.repeatable.JoinSpecifications
import io.micronaut.data.jdbc.annotation.JdbcRepository
import io.micronaut.data.model.query.builder.sql.Dialect
import io.micronaut.data.repository.kotlin.KotlinCrudRepository

@JdbcRepository(dialect = Dialect.H2)
interface SaleRepository : KotlinCrudRepository<Sale, Long> {

    @JoinSpecifications(value = [Join("product"), Join("product.manufacturer")])
    override fun findById(id: Long): Sale?
}

File: doc-examples/jdbc-example-python/src/main/python/example/SaleRepository.py
from java.util import Optional
from micronaut.data.annotation import Join
from micronaut.data.jdbc.annotation import JdbcRepository
from micronaut.data.model.query.builder.sql import Dialect
from micronaut.data.repository import CrudRepository

from example.Quantity import Quantity
from example.Sale import Sale


@JdbcRepository(dialect=Dialect.H2)
class SaleRepository(CrudRepository[Sale, int]):

    @Join("product")
    @Join("product.manufacturer")
    def getById(self, id: int) -> Optional[Sale]: ...  # TODO(python): overriding the inherited findById(Integer) drops the method

    @Join("product")
    @Join("product.manufacturer")
    def findByQuantity(self, quantity: Quantity) -> Optional[Sale]: ...
